monad-rpc/src/handlers/eth/call.rs

```diff
@@ -1002,6 +1006,220 @@ pub async fn monad_admin_ethCallStatistics(
     })
 }
 
+#[derive(Debug, Deserialize, schemars::JsonSchema)]
+pub struct MonadEthFillTransactionParams {
+    pub tx: CallRequest,
+}
+
+#[rpc(
+    method = "eth_fillTransaction",
+    ignore = "chain_id,provider_gas_limit,eth_call_executor"
+)]
+#[allow(non_snake_case)]
+pub async fn monad_eth_fillTransaction<T: Triedb>(
+    triedb_env: &T,
+    eth_call_executor: Arc<EthCallExecutor>,
+    chain_id: u64,
+    provider_gas_limit: u64,
+    params: MonadEthFillTransactionParams,
+) -> JsonRpcResult<FillTransactionResult> {
+    trace!("monad_eth_fillTransaction: {params:?}");
+
+    let mut tx = params.tx;
+
+    tx.input.input = match (tx.input.input.take(), tx.input.data.take()) {
+        (Some(input), Some(data)) => {
+            if input != data {
+                return Err(JsonRpcError::invalid_params());
+            }
+            Some(input)
+        }
+        (None, data) | (data, None) => data,
+    };
+
+    let from = tx.from.ok_or_else(JsonRpcError::invalid_params)?;
+
+    if tx.value.is_none() {
+        tx.value = Some(U256::from(0));
+    }
+
+    tx.chain_id = Some(U64::from(chain_id));
+
+    let block_key = get_block_key_from_tag(triedb_env, BlockTags::Latest)
+        .ok_or(JsonRpcError::block_not_found())?;
+
+    if tx.nonce.is_none() {
+        let account = triedb_env
+            .get_account(block_key, from.into())
+            .await
+            .map_err(JsonRpcError::internal_error)?;
+        tx.nonce = Some(U64::from(account.nonce));
+    }
+
+    let header = triedb_env
+        .get_block_header(block_key)
+        .await
+        .map_err(JsonRpcError::internal_error)?
+        .ok_or_else(JsonRpcError::block_not_found)?;
+
+    // Fill max_priority_fee_per_gas with suggested fee if not specified (EIP-1559 only)
+    if let GasPriceDetails::Eip1559 {
+        max_fee_per_gas,
+        max_priority_fee_per_gas: None,
+    } = &tx.gas_price_details
+    {
+        let priority_fee = suggested_priority_fee().await.unwrap_or_default();
+        tx.gas_price_details = GasPriceDetails::Eip1559 {
+            max_fee_per_gas: *max_fee_per_gas,
+            max_priority_fee_per_gas: Some(U256::from(priority_fee)),
+        };
+    }
+
+    // Fill gas prices (max_fee_per_gas, max_priority_fee_per_gas, or gas_price)
+    let base_fee = U256::from(header.header.base_fee_per_gas.unwrap_or_default());
+    tx.fill_gas_prices(base_fee)?;
+
+    // Estimate gas if not specified
+    let gas_specified = tx.gas.is_some();
+    if !gas_specified {
+        let protocol_gas_limit = header.header.gas_limit;
+        let eth_call_provider_gas_limit = provider_gas_limit.min(protocol_gas_limit);
+        let original_tx_gas = U256::from(protocol_gas_limit);
+
+        // Set initial gas for estimation
+        tx.gas = Some(U256::from(eth_call_provider_gas_limit));
```

Comment:
Do we want to set the initial gas limit  to the provider limit or use `fill_gas_params`? That method will fill both prices and the gas limit to the sender's actual gas allowance based on their balance.